tests/basic.rs

```diff
@@ -1,486 +1,860 @@
+                .get_entry_follow(path)
+                .ok_or_else(|| Error::new(ErrorKind::NotFound, "metadata: entry not found"))?;
+
+            match &entry {
+                DirectoryEntry::Directory(dir) => {
+                    let entries = dir
+                        .entries
+                        .keys()
+                        .map(|name| {
+                            Ok(ReadDirEntry {
+                                path: path.join(name),
+                            })
+                        })
+                        .collect::<Vec<_>>();
+                    Ok(Box::new(entries.into_iter()))
+                }
+                // should use ErrorKind::NotADirectory once upgrading rust version
+                _ => Err(Error::new(ErrorKind::Other, "Not a directory")),
+            }
+        }
+
+        fn is_valid_executable(&self, path: &Path) -> std::io::Result<bool> {
+            let entry = self.get_entry_follow(path).ok_or_else(|| {
+                Error::new(ErrorKind::NotFound, "is_valid_executable: entry not found")
+            })?;
+
+            match &entry {
+                DirectoryEntry::File(file) => Ok(file.is_valid_executable),
+                _ => Ok(false),
+            }
+        }
+    }
+
+    #[test]
+    fn basic() {
+        let mut sys = InMemorySys::new();
+        sys.set_env_var("PATH", "/sub/dir1/:/sub/dir2/");
+        sys.write_non_executable("/sub/dir1/exec1");
+        sys.write_executable("/sub/dir2/exec1"); // will get this one
+        sys.write_executable("/sub/dir2/exec2");
+        let config = which::WhichConfig::new_with_sys(sys).binary_name(OsString::from("exec1"));
+        let result = config.first_result().unwrap();
+        assert_eq!(result, PathBuf::from("/sub/dir2/exec1"));
+    }
 
-#[test]
-#[cfg(windows)]
-fn windows_no_extension_but_executable() {
-    let this_executable = std::env::current_exe().unwrap();
-    let new_name = this_executable.parent().unwrap().join("test_executable");
-    std::fs::copy(&this_executable, &new_name).unwrap();
-    let found_executable = which::which_in_global(
-        new_name.file_name().unwrap(),
-        Some(this_executable.parent().unwrap()),
-    )
-    .unwrap()
-    .next()
-    .unwrap();
-    assert_eq!(found_executable, new_name);
-    std::fs::remove_file(new_name).unwrap();
+    #[test]
+    fn symlink() {
+        let mut sys = InMemorySys::new();
+        sys.set_env_var("PATH", "/sub/dir1/");
+        sys.create_symlink("/sub/dir1/exec", "/sub/dir2/exec");
+        sys.write_executable("/sub/dir2/exec");
+        let config = which::WhichConfig::new_with_sys(sys).binary_name(OsString::from("exec"));
+        let result = config.first_result().unwrap();
+        assert_eq!(result, PathBuf::from("/sub/dir1/exec"));
+    }
+
+    #[test]
+    fn tilde_path() {
+        let mut sys = InMemorySys::new();
+        sys.set_home_dir("/home/user/");
+        sys.set_env_var("PATH", "/dir/:~/sub/");
```

Comment:
I'm actually surprised this crate has this behaviour, but it's nice it can be disabled now by overriding the `home_dir` implementation to return `None`.